### PR TITLE
Refactor RepairQueue logic into ReplicateQueue

### PR DESCRIPTION
**Pull request by mrtracy**
A recent commit created the RepairQueue, a queue responsible for detecting dead
replicas. This commit moves dead replica detection and removal to the
ReplicateQueue, removing the RepairQueue in the process.

### CODE REVIEW

File: storage/allocator_test.go [hunk 1]
```diff
@@ -173,6 +173,27 @@ func createTestAllocator() (*stop.Stopper, *gossip.Gossip, *StorePool, allocator
...
 
+// mockStorePool sets up a collection of a alive and dead stores in the
+// store pool for testing purposes.
+func mockStorePool(storePool *StorePool, aliveStoreIDs, deadStoreIDs []proto.StoreID) {
+	storePool.mu.Lock()
+	defer storePool.mu.Unlock()
+
+	storePool.stores = make(map[proto.StoreID]*storeDetail)
+	for _, storeID := range aliveStoreIDs {
+		storePool.stores[storeID] = &storeDetail{
```
Comment (tamird) [inline storage/allocator_test.go]: it's not obvious to me why this map has pointer values; can you see about changing them to value types? if that's not possible, could you add a comment on storePool?

File: storage/store_pool.go [hunk 2]
```diff
@@ -259,6 +259,18 @@ func (sp *StorePool) getStoreDescriptor(storeID proto.StoreID) *proto.StoreDescr
 	return &desc
 }
 
+// findDeadReplicas returns any replicas from the supplied slice that are
+// located on dead stores.
+func (sp *StorePool) findDeadReplicas(repls []proto.Replica) []proto.Replica {
```
Comment (tamird) [inline storage/store_pool.go]: s/findD/d/

File: storage/store_pool_test.go [hunk 3]
```diff
@@ -294,3 +294,78 @@ func TestStorePoolGetStoreDetails(t *testing.T) {
...
+	}
+	// Timeout all stores, but specifically store 5.
+	waitUntilDead(t, sp, 5)
+
+	// Resurrect all stores except for 4 and 5.
+	sg.GossipStores(stores[:3], t)
+
+	deadReplicas = sp.findDeadReplicas(replicas)
+	if a, e := deadReplicas, replicas[3:]; !reflect.DeepEqual(a, e) {
+		t.Fatalf("findDeadReplicas did not return expected values; got %v, expected %v", a, e)
```
Comment (tamird) [inline storage/store_pool_test.go]: suggestion: print a newline before printing a slice

**Comment by tamird** [comment]
LGTM looks like mostly code movement

File: storage/repair_queue_test.go [hunk 4]
```diff
@@ -1,155 +0,0 @@
...
-		// Test 4: five store, one dead stores, repair the store
-		{
-			replica:        createTestReplicaForRepair([]proto.StoreID{1, 2, 3, 4, 5}),
-			storePoolAlive: []proto.StoreID{1, 2, 3, 4},
-			storePoolDead:  []proto.StoreID{5},
-			eNeeds:         true,
-			ePriority:      -1,
-			eReplicas:      createReplicaSets([]proto.StoreID{5}),
-		},
-		// Test 5: three stores, all alive, but only 1 is in the store pool, nothing to do
```
Comment (BramGruneir) [inline storage/repair_queue_test.go]: Can you find a way to test out this case?  Where there is no info yet in the store pool.
This is a weird case, but we give the benefit of not knowing if a store even exists as it being alive (and start the timer right there)

Comment (BramGruneir) [inline storage/repair_queue_test.go]: Also test cases without quorum and when more than one store is dead, but quorum still exists.

**Comment by BramGruneir** [comment]
LGTM with extra test cases.

File: storage/repair_queue_test.go (hunk 4 above)
Comment (mrtracy) [inline storage/repair_queue_test.go]: Added a few test cases: multiple dead nodes, nodes which aren't in the storePool.

File: storage/store_pool.go (hunk 2 above)
Comment (mrtracy) [inline storage/store_pool.go]: Done

File: storage/store_pool_test.go (hunk 3 above)
Comment (mrtracy) [inline storage/store_pool_test.go]: Done

File: storage/allocator_test.go (hunk 1 above)
Comment (mrtracy) [inline storage/allocator_test.go]: Done

File: storage/store_pool.go [hunk 5]
```diff
@@ -259,6 +261,18 @@ func (sp *StorePool) getStoreDescriptor(storeID proto.StoreID) *proto.StoreDescr
 	return &desc
 }
 
+// findDeadReplicas returns any replicas from the supplied slice that are
```
Comment (spencerkimball) [inline storage/store_pool.go]: The comments name this `findDeadReplicas`, but the method is actually called `deadReplicas`. I think `findDeadReplicas` is actually a better name as it contains a verb.

**Pull request merged**